automerge/src/combined.rs

```diff
@@ -0,0 +1,247 @@
+use std::collections::HashMap;
+
+use automerge_backend::{Backend, Change, SyncMessage, SyncState};
+use automerge_frontend::{value_ref::RootRef, Frontend, MutableDocument, Path, Value};
+use automerge_protocol as amp;
+use automerge_protocol::OpId;
+use thiserror::Error;
+
+use crate::{BackendError, FrontendError, InvalidPatch};
+
+#[derive(Debug, Error)]
+pub enum AutomergeError {
+    #[error(transparent)]
+    BackendError(#[from] BackendError),
+    #[error(transparent)]
+    FrontendError(#[from] FrontendError),
+    #[error(transparent)]
+    InvalidPatch(#[from] InvalidPatch),
+    #[error(transparent)]
+    UserError(#[from] Box<dyn std::error::Error>),
+}
+
+/// The core automerge document.
+///
+/// In reality this combines both the [`Frontend`] and [`Backend`] and handles synchronising the frontend
+/// with updates from the backend caused by local changes.
+#[derive(Debug, Default)]
+pub struct Automerge {
+    frontend: Frontend,
+    backend: Backend,
+}
+
+impl Automerge {
+    /// Construct a new, empty, automerge instance.
+    ///
+    /// To build a new document from a previously saved one see [`load`](Self::load).
+    pub fn new() -> Self {
+        Self {
+            frontend: Frontend::new(),
+            backend: Backend::new(),
+        }
+    }
+
+    /// Get the current value that this document stores.
+    pub fn state(&mut self) -> &Value {
+        self.frontend.state()
+    }
+
+    /// Get a reference to the value that this document stores, without building it all.
+    pub fn value_ref(&self) -> RootRef {
+        self.frontend.value_ref()
+    }
+
+    /// Make a change to this document.
+    ///
+    /// Internally, this applies the updates to the frontend to get a change.
+    /// This change is then applied to the backend in return for a patch.
+    /// This patch is then used to update the frontend's bookkeeping.
+    pub fn change<F, O, E>(
+        &mut self,
+        message: Option<String>,
+        change_closure: F,
+    ) -> Result<(O, Option<&Change>), AutomergeError>
+    where
+        E: std::error::Error + 'static,
+        F: FnOnce(&mut dyn MutableDocument) -> Result<O, E>,
+    {
+        let (out, change) = self
+            .frontend
+            .change(message, change_closure)
+            .map_err(|e| AutomergeError::UserError(Box::new(e)))?;
```

Comment:
Do we need to return errors here and in the following `apply_patch`. The change has been generated by the frontend we control and likewise the patch is generated by the backend we control and delivered in order. I think there is no way this can fail which isn't a programming error on our part, in which case we should probably just panic?

Reply:
Agreed, thanks for the simplification. I've removed the variant on the `AutomergeError` enum too now.